### DISCUSSION TITLE
[FEATURE] Enhance Bokeh Tooltips for Richer Data Context

### DISCUSSION THREAD

**Discussion by ChrisBSK** [bokeh/bokeh#14166] [Internals and design (Q&A)]
### Problem description

Currently, Bokeh tooltips are limited to showing basic information about the data points (e.g., X and Y values). This works well for simple visualizations, but it can be insufficient when trying to provide more detailed or context-rich information. For example, in bar charts or other plots with categorical data, users often need more context such as percentages relative to the total, additional commentary, or other statistics to better understand the data.

In this following example, we would like to have more precise information on the points shown on the plot, than only the type of penguin:
![Image](https://github.com/user-attachments/assets/88c35a77-1957-4ba2-b0bd-cb04f45f661b)

Without a way to dynamically include additional information in the tooltip, users might miss out on key insights or have to rely on external sources or complex annotations outside the plot.

### Feature description

We propose enhancing the current tooltip functionality by allowing dynamic tooltips that display more comprehensive information. This feature would allow the tooltip to show:

-  **Basic information** such as X and Y values.

- **Derived statistics** such as the percentage of each value relative to the total.

- **Contextual comments or insights**, such as annotations or labels that describe the importance of a given data point.

The tooltip should update dynamically based on the data being hovered over, displaying additional context without cluttering the plot itself.

Additionally, this feature should allow for easy customization of the tooltip content, enabling users to specify what information they would like to show.

### Potential alternatives

I've been working on this and I have a proposition about some new arguments and their behavior. I will post a pull request so we can discuss if the solution is something of interest to the community

I've been working on this and I have a proposition about some new arguments and their behavior. I will post a pull request so we can discuss if the solution is something of interest to the community but to give you an intuition it will look like this

**Solution**

When the user hovers over a data point (such as a circle or a bar), the tooltip displays additional information like the exact value, as well as contextual information that changes based on the categories or the axis.

```python
from bokeh.plotting import figure, show
from bokeh.sampledata.penguins import data
from bokeh.transform import factor_cmap, factor_mark
from bokeh.models import HoverTool, ColumnDataSource
import pandas as pd

# Preprocessing for hover details
SPECIES = sorted(data.species.unique())
MARKERS = ['hex', 'circle_x', 'triangle']

# Calculate percentage of each species in the total population
species_counts = data['species'].value_counts()
data['species_percentage'] = data['species'].map(
    lambda x: f"{(species_counts[x] / len(data) * 100):.2f}%"
)

# Define adaptability based on body mass (in g)
data['adaptability'] = pd.cut(data['body_mass_g'], 
                              bins=[0, 3500, 4500, 6000], 
                              labels=["+", "++", "+++"])

source = ColumnDataSource(data)

# Create figure
p = figure(title="Penguin Size by Species", background_fill_color="#fafafa")
p.xaxis.axis_label = 'Flipper Length (mm)'
p.yaxis.axis_label = 'Body Mass (g)'

# Scatter plot with colors and markers
p.scatter("flipper_length_mm", "body_mass_g", source=source,
          legend_group="species", fill_alpha=0.4, size=12,
          marker=factor_mark('species', MARKERS, SPECIES),
          color=factor_cmap('species', 'Category10_3', SPECIES))

# Add hover tool
hover = HoverTool()
hover.tooltips = [
    ("Species", "@species"),
    ("Species Percentage", "@species_percentage"),
    ("Flipper Length", "@flipper_length_mm mm"),
    ("Body Mass", "@body_mass_g g"),
    ("Adaptability", "@adaptability")
]
p.add_tools(hover)

# Configure legend
p.legend.location = "top_left"
p.legend.title = "Species"

# Show plot
show(p)

```

![Image](https://github.com/user-attachments/assets/fb3d9e45-0256-4794-a302-0878069fe59c)
![Image](https://github.com/user-attachments/assets/6f99dc6d-4316-4a3e-8dfa-e1a4d527c49a)
![Image](https://github.com/user-attachments/assets/83c7f961-2d7f-4d72-8095-42059e166915)

### Additional information

- This feature would greatly enhance the usability of Bokeh for more complex data visualizations, such as financial or business intelligence dashboards, where context and insights are essential.

- A similar feature exists in libraries like **Plotly** and **Matplotlib**, where hover interactions display enriched information.

- This feature should work with existing Bokeh plotting capabilities, maintaining compatibility with other tools like legends, axis labels, and interaction controls.

- Customizable tooltips would also allow users to decide what information is most relevant for their visualizations, improving the adaptability of Bokeh.

It would be a pleasure to contribute to this feature if it's accepted, I really like the mindset of Bokeh's Community. 

@ChrisBSKK.

**Comment by bryevdv** [comment]
@ChrisBSK I have to admit I am confused by this issue. All of the things you list:

> Basic information such as X and Y values.
> Derived statistics such as the percentage of each value relative to the total.
> Contextual comments or insights, such as annotations or labels that describe the importance of a given data point.

already exist in Bokeh, and have for a very long time. There are already tooltip fields for positions (screen or data, numerical or categorical or datetime), for derived and contextual values (either stored ahead of time in a CDS or computed "on the fly" with a `CustomJSHover`)

 I think what would help the most, and immediately clarify things, would be to provide a mock-up (or reference) to an example plot you want to be able to make, but can't currently, due to a limitation with Bokeh. Alternatively a proof-of-concept PR could help focus things. 

Edit: I just noticed this sentence

> I've been working on this and I have a proposition about some new arguments and their behavior. I will post a pull request so we can discuss if the solution is something of interest to the community

Ideally, these kind of details, i.e. concrete suggestions for functions or arguments, are what go the feature request itself.

**Comment by ChrisBSK** [comment]
Thank you very much @bryevdv for your answer ! 

I didn't notice that there were that much functionalities already implemented in Bokeh, I learned a lot. 

I have thought about another situation which came to my mind when I saw this stack overflow sollicitation: https://stackoverflow.com/questions/48785310/having-broken-y-axis-in-bokeh-plot-figure

### Problem description

Bokeh does not have native support for **broken axis plots**, which are essential for visualizing data with values spanning different scales. In such cases, a traditional axis can either compress one dataset or fail to effectively represent values across both datasets. For example, when plotting both **exponential growth** (which grows rapidly) and **quadratic growth** (which grows more slowly), a single axis becomes unsuitable as it distort the view of one dataset while trying to fit the other.

In the following example, we are dealing with two datasets, one showing exponential growth (y = exp(x)) and another showing quadratic growth (y = x²). Without a broken axis, the exponential data would dominate the plot, making it difficult to visualize both datasets clearly:

![Image](https://github.com/user-attachments/assets/667f0817-ec13-46ce-9f35-79561fb31f52)

### Feature description

I propose adding a feature for **broken axis plots** in Bokeh, if It does not already exist. This feature would allow users to create plots where the y-axis (or x-axis) is split into two or more segments, each displaying a different range of data. The break in the axis visually indicates that part of the axis is intentionally omitted, making the plot easier to interpret when dealing with data of significantly different scales.

Key points of the broken axis feature:
- Users can define which sections of the axis should be broken (e.g., for displaying large exponential growth next to quadratic growth).
- The plot would visually show the axis break providing clarity without losing any important data points.
- The tooltips should adapt to the different axis segments, offering context-sensitive information depending on the section the user is hovering over.

### Proposed solution

We propose the following solution to implement broken axis plots in Bokeh:

- The y-axis will be divided into two parts: one for the higher range of data (exponential growth) and one for the lower range (quadratic growth).
- The axis break will be visually represented with diagonal lines.
- The user will be able to hover over the data points, and the tooltips will dynamically adjust based on the axis segment.

Implemented in **Matplotlib** it looks like this (Can looks better depending on the parameters used):
![Image](https://github.com/user-attachments/assets/f96d1e37-e3b5-4924-b75a-cb719aec27f8)
With the following code:
```python
import matplotlib.pyplot as plt
import numpy as np

# Generate simulated data
x = np.linspace(0, 10, 500)
y1 = np.exp(x)  # Exponential growth
y2 = x**2       # Quadratic growth

# Create figure with two axes (broken axis)
fig, (ax1, ax2) = plt.subplots(2, 1, sharex=True, figsize=(8, 6), gridspec_kw={'height_ratios': [1, 1]})

# Plot the data
ax1.plot(x, y1, color="blue", label="Exp Growth (y=exp(x))")
ax2.plot(x, y2, color="orange", label="Quadratic Growth (y=x²)")

# Set limits for each axis
ax1.set_ylim(10**2, 10**4)  # Upper range (log-like scale)
ax2.set_ylim(0, 100)        # Lower range (zoomed-in)

# Hide spines to indicate broken axis
ax1.spines['bottom'].set_visible(False)
ax2.spines['top'].set_visible(False)
ax1.xaxis.tick_top()
ax1.tick_params(labeltop=False)  # No labels on the top axis
ax2.xaxis.tick_bottom()

# Add diagonal lines to represent the break
d = .015  # Diagonal line size
kwargs = dict(transform=ax1.transAxes, color='k', clip_on=False)
ax1.plot((-d, +d), (-d, +d), **kwargs)  # Top-left diagonal
ax1.plot((1 - d, 1 + d), (-d, +d), **kwargs)  # Top-right diagonal

kwargs.update(transform=ax2.transAxes)
ax2.plot((-d, +d), (1 - d, 1 + d), **kwargs)  # Bottom-left diagonal
ax2.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)  # Bottom-right diagonal

# Titles, labels, and legend
ax2.set_xlabel("X-axis")
ax1.set_ylabel("Y-axis (high range)")
ax2.set_ylabel("Y-axis (low range)")

ax1.legend(loc="upper left")
ax2.legend(loc="upper left")
plt.suptitle("Broken Axis Plot Example")

# Display plot
plt.tight_layout()
plt.show()
```
But the idea is to transpose that to **Bokeh**.

Below is a first insight of how I imagine the plots in Bokeh:

```python
from bokeh.plotting import figure, show
from bokeh.models import HoverTool, ColumnDataSource
from bokeh.layouts import column
from bokeh.io import output_file
import numpy as np

# Generate simulated data
x = np.linspace(0, 10, 500)
y1 = np.exp(x)  # High range data (exponential growth)
y2 = x**2       # Low range data (quadratic growth)

# Create a ColumnDataSource
source = ColumnDataSource(data=dict(x=x, y1=y1, y2=y2))

# Create two separate figures (one for each plot)
p1 = figure(title="Exponential Growth", background_fill_color="#fafafa", tools="pan,box_zoom,reset,save", height=300)
p2 = figure(title="Quadratic Growth", background_fill_color="#fafafa", tools="pan,box_zoom,reset,save", height=300, x_range=p1.x_range)

# Plot the exponential growth
p1.line('x', 'y1', source=source, line_width=2, color="blue", legend_label="Exp Growth (y=exp(x))")

# Plot the quadratic growth
p2.line('x', 'y2', source=source, line_width=2, color="orange", legend_label="Quadratic Growth (y=x²)")

# Add hover tool to both plots
hover = HoverTool()
hover.tooltips = [
    ("X", "@x"),
    ("Exponential Growth (y=exp(x))", "@y1"),
    ("Quadratic Growth (y=x²)", "@y2"),
]
p1.add_tools(hover)
p2.add_tools(hover)

# Set axis labels
p1.xaxis.axis_label = 'X-axis'
p2.xaxis.axis_label = 'X-axis'

p1.yaxis.axis_label = 'Exponential Y-axis'
p2.yaxis.axis_label = 'Quadratic Y-axis'

# Set y-limits for each plot to simulate broken axis
p1.y_range.start = 100
p1.y_range.end = 10000  # Focus on the exponential part
p2.y_range.start = 0
p2.y_range.end = 100  # Focus on the quadratic part

# Remove the top and bottom spines to simulate break
p1.xaxis.visible = False  # Hide x-axis labels and ticks for p1
p2.xaxis.visible = True  # Show x-axis labels and ticks for p2

# Add diagonal lines to simulate break (visual effect)
d = .02  # Size of the diagonal lines (adjustable)
kwargs = dict(color='black', line_width=2)

# Draw diagonal lines at the "break" between the two plots
p1.line([0, 1], [1-d, 1-d], **kwargs)  # Top line for p1
p2.line([0, 1], [0+d, 0+d], **kwargs)  # Bottom line for p2

# Add the plot to a layout
layout = column(p1, p2)

# Show the plot
output_file("broken_axis_plot.html")  # Output to an HTML file
show(layout)
```
Giving the following plot for the moment:
![Image](https://github.com/user-attachments/assets/9c837506-419d-4d78-a1f3-2c63c83ef9e3)

I think it can be a very useful feature if it doesn't already exist. 

Thank you in advance

@ChrisBSK.

**Comment by bryevdv** [comment]
Hi @ChrisBSK issues really need to be single-focus. As you have veered rather sharply onto a different unrelated topic, I have gone ahead and converted this to a discussion instead. You seem to interested to find a suitable target for contributing (which is great!) but we can open a new focused issue once something preliminary is hashed out here.


You are correct that Bokeh does not currently support broken axes directly. I believe there was a previous very old issue regarding adding support for them, but I have not been able to find it after an initial search. I will start out by saying there are definitely some thigns that need to be thought through up front because the answers are not clear cut. 

First off. I can imagine two alternatives:

## One single actually broken axis on a single plot

This is more like the MPL case you show above. I will say that I think this would be very difficult in Bokeh. Specifically, the MPL plot is static and does not have to support interactions like panning and zooming, etc. MPL can just special-case some drawing for broken axes and call it a day, since no interactions will trigger any updates of ay kind. 

This would make "broken axes" an actual first-class feature directly inside Bokeh. If this is the scenario you actually imagine for Bokeh then you will need to think very hard about *exactly* how broken axes will intersect with all of Bokeh's interactive capabilities. Additionally, note that this will require new BokehJS objects like some kind of "broken" range mapper, and updates to some of the complex Axis code to deal with them, possibly simples with an entirely new "BrokenAxis" type to go along.

## "Faking" broken axes with two plots

This is like your Bokeh example code above. More or less, nothing new is *required* in terms of implementation, API, etc. I think it could  even be argued that adding some new examples and a section in the users guide would even be sufficient to be "done". However, I can also imagine some Python-side convenience API that helps with setting up the multiple plots more easily, with less boilerplate code. 

This would make "broken axes" less of a feature in and of itself, and more like just a technique to be demonstrated and documented. This approach would be *much* easier to do, for sure. 

---

So the first question is: which of those possibilities do you actually have in mind?

**Reply by ChrisBSK** [reply]
Thank you @bryevdv, I understand much better how MPL handles those things now and as you said I thing that will be very difficult to implement it with all those parameters to take into account. 

That's clearly the second possibility that I had in mind when I wrote this situation, something more practical. 

In this possibility, I thing that it can be a very good idea to add some new examples and a section in the users guide as you said in a first time. 
And maybe later go deeper and provide some Python-side convenience API, in a second phase.

@ChrisBSK.
